src/pie_utils/processor/document/candidate_relation_adder.py

```diff
@@ -0,0 +1,211 @@
+from __future__ import annotations
+
+import json
+import logging
+import random
+from collections import defaultdict
+from typing import Any, Dict, List
+
+from pytorch_ie.annotations import BinaryRelation, Span
+from pytorch_ie.utils.span import is_contained_in
+
+from pie_utils.span.slice import distance
+from pie_utils.statistics import WithStatistics
+
+from ..document import DocumentWithEntitiesRelationsAndPartition
+
+logger = logging.getLogger(__name__)
+
+
+class CandidateRelationAdder(WithStatistics):
+    """CandidateRelationAdder adds binary relations to a document based on various parameters. It
+    goes through combinations of available entity pairs as possible candidates for new relations.
+    Entity pairs which are already part of document as a relation are not added again.
+
+    params:
+        label: label for the new relations to be added
+        use_partition: A boolean parameter to enable partition wise relation creation. If this parameter is enabled then
+                    it uses available partitions in the document otherwise use whole document as single partition. Note
+                    that entity pairs which belong to the same partition are considered for the creation of new relations.
+        max_distance: An optional parameter which restricts the maximum distance between candidate entity pair to form a
+                    new relation. If distance between entity pair is more than the value of this parameter then candidate
+                    pair is discarded.
+        distance_type: type of distance to be calculated between two entities, it can be inner, outer or center. It is
+                    important to note that value for max_distance parameter should be given keeping value of this parameter
+                    in mind.
+        added_rels_upper_bound_factor: It is an optional parameter used to calculate the upper bound for the number of
+                                    new relations that can be added. Upper bound is calculated as,
+                                                n = k * num_available_relations
+                                    where n is upper bound, k is added_rels_upper_bound_factor and see collect_statistics
```

Comment:
```suggestion
        added_rels_upper_bound_factor: [DEPRECATED] It is an optional float parameter to restrict the number of added relations by an upper bound. The upper bound n is calculated as,
                                                n = added_rels_upper_bound_factor * num_available_relations
                                    see collect_statistics for the definition of num_available_relations.
                                    Note: This is deprecated and should not be used because because during testing, the num_available_relations should not be available. Use max_distance instead which does not depend on that.
```